### DISCUSSION TITLE
How to use `distance_in_km_level` function?

### DISCUSSION THREAD

**Discussion by matthewfeldman** [moj-analytical-services/splink#1038] [General]
### What happens?

I am trying to find in the documentation an example of using the haversine distance functions for setting comparison levels. 

I am looking to use something like 
```
settings = {
    "link_type": "dedupe_only",
    "blocking_rules_to_generate_predictions": [
        "l.gh6 = r.gh6",
        "l.address = r.address",
        "l.name = r.name"
    ],
    "comparisons": [
        levenshtein_at_thresholds("name", [5,10,20],term_frequency_adjustments=True),
        levenshtein_at_thresholds("address",[2,5,10],term_frequency_adjustments=True),
        levenshtein_at_thresholds('geohash',[1,3,5,7]),
        distance_in_km_level(long_col='lng',lat_col='lat',km_threshold=[0.01,0.1]),
        
    ],
}
```

But I am getting this error
File /code/venvs/venv/lib/python3.8/site-packages/splink/settings.py:39, in Settings.__init__(self, settings_dict)
     35 # In incoming comparisons have nested ComparisonLevels, turn back into dict
     36 for comparison_dict in settings_dict["comparisons"]:
     37     comparison_dict["comparison_levels"] = [
     38         cl.as_dict() if isinstance(cl, ComparisonLevel) else cl
---> 39         for cl in comparison_dict["comparison_levels"]
     40     ]
     42 validate_settings_against_schema(settings_dict)
     44 self._settings_dict = settings_dict

TypeError: 'distance_in_km_level' object is not subscriptable

### To Reproduce

from splink.duckdb.duckdb_linker import DuckDBLinker

from splink.duckdb.duckdb_comparison_library import (
    exact_match,
    levenshtein_at_thresholds,
    distance_function_at_thresholds,
)
from splink.duckdb.duckdb_comparison_level_library import distance_in_km_level

settings = {
    "link_type": "dedupe_only",
    "blocking_rules_to_generate_predictions": [
        "l.gh6 = r.gh6",
        "l.address = r.address",
        "l.name = r.name"
    ],
    "comparisons": [
        levenshtein_at_thresholds("name", [5,10,20],term_frequency_adjustments=True),
        levenshtein_at_thresholds("address",[2,5,10],term_frequency_adjustments=True),
        levenshtein_at_thresholds('geohash',[1,3,5,7]),
        distance_in_km_level(long_col='lng',lat_col='lat',km_threshold=0.01),
        
    ],
}

linker = DuckDBLinker(df_places[['unique_id','name','geohash','gh6','address','lat','lng','name_simple']], settings)
linker.estimate_u_using_random_sampling(target_rows=1e6)

blocking_rule_for_training = "l.address = r.address"
linker.estimate_parameters_using_expectation_maximisation(blocking_rule_for_training)

pairwise_predictions = linker.predict()

clusters = linker.cluster_pairwise_predictions_at_threshold(pairwise_predictions, 0.85)
clusters.as_pandas_dataframe(limit=5)

### OS:

python 3.8

### Splink version:

3.6.0

### Have you tried this on the latest `master` branch?

- [x] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [ ] I agree

**Comment by ADBond** [comment]
The issue here is that `distance_in_km_level` is a single _comparison level_. The comparisons that you include in a settings dictionary need to be made up of a variety of comparison levels - see [the docs page on this](https://moj-analytical-services.github.io/splink/topic_guides/customising_comparisons.html), in particular [the section on using custom comparison levels](https://moj-analytical-services.github.io/splink/topic_guides/customising_comparisons.html#method-2-comparisonlevels).

In your case you would need something like:
```py
from splink.duckdb.duckdb_comparison_level_library import (
    null_level,
    else_level,
    distance_in_km_level,
)
...
"comparisons": [
    levenshtein_at_thresholds("name", [5,10,20],term_frequency_adjustments=True),
    levenshtein_at_thresholds("address",[2,5,10],term_frequency_adjustments=True),
    levenshtein_at_thresholds('geohash',[1,3,5,7]),
    {
        'output_column_name': 'lat_long_distance',
        'comparison_description': 'Lat-long distance at 0.01, 0.1 km vs anything else',
        'comparison_levels': [
            # always need to include a null level, for situations where a comparison doesn't really make sense
            {
                "sql_condition": "(lat_l IS NULL OR lat_r IS NULL) OR (long_l IS NULL OR long_r IS NULL)",
                "label_for_charts": "Null",
                "is_null_level": True,
            },
            distance_in_km_level(long_col='lng', lat_col='lat', km_threshold=0.01),
            distance_in_km_level(long_col='lng', lat_col='lat', km_threshold=0.1),
            # also always need an else level to capture any comparisons not covered by the other levels
            else_level()
        ],
    },

],
```

The reason we have to include the `null` level via a dictionary is due to the fact we need to check if either `lat` or `lng` column is `NULL` - this will get easier after we have implemented #885.

I also had not realised that we do not have a ready-made `comparison` that does this automatically - apologies, that is an oversight on my part! I will raise an issue for that as well, so that in future you will be able to use code more similar to what you originally had.

**Reply by ADBond** [reply]
Have added a feature request for a ready-made comparison at a variety of user-supplied thresholds #1039

**Reply by matthewfeldman** [reply]
Oh my gosh you are so fast!  Will take a look at this approach and seeing if I can get it running now!

**Comment by RossKen** [comment]
@matthewfeldman we have just merged in a `cl.distance_in_km_at _thresholds` function that will be part of our next release, see #1117 for more details  😊
